system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <assert.h>
+#include <emscripten/threading.h>
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "proxying.h"
+
+#define TASK_QUEUE_INITIAL_CAPACITY 128
+
+extern int _emscripten_notify_proxying_queue(pthread_t target_thread,
+                                             pthread_t curr_thread,
+                                             pthread_t main_thread,
+                                             em_proxying_queue* queue);
+
+typedef struct task {
+  void (*func)(void*);
+  void* arg;
+} task;
+
+// A task queue for a particular thread. Organized into a linked list of
+// task_queues for different threads.
+typedef struct task_queue {
+  // The target thread for this task_queue.
+  pthread_t thread;
+  // Recursion guard. TODO: We disallow recursive processing because that's what
+  // the old proxying API does. Experiment with relaxing this restriction.
+  int processing;
+  // Ring buffer of tasks of size `capacity`. New tasks are enqueued at
+  // `tail` and dequeued at `head`.
+  task* tasks;
+  int capacity;
+  int head;
+  int tail;
+} task_queue;
+
+static int task_queue_init(task_queue* tasks, pthread_t thread) {
+  task* task_buffer = malloc(sizeof(task) * TASK_QUEUE_INITIAL_CAPACITY);
+  if (task_buffer == NULL) {
+    return 0;
+  }
+  *tasks = (task_queue){.thread = thread,
+                        .processing = 0,
+                        .tasks = task_buffer,
+                        .capacity = TASK_QUEUE_INITIAL_CAPACITY,
+                        .head = 0,
+                        .tail = 0};
+  return 1;
+}
+
+static void task_queue_deinit(task_queue* tasks) { free(tasks->tasks); }
```

Comment:
`task_queue_free()`? (feel free to ignore suggestion)

Reply:
I would like to keep `deinit` because it is more clearly pairs with `init` and because it's not `tasks` that's being freed, but rather one of its members.